lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -171,9 +172,273 @@ bool swift::isIntermediateRelease(SILInstruction *inst,
+///   (2) The call may destroy owned parameters, which are guaranteed to have
+///   value semantics as the call is constant evaluable. (3) The call does not
+///   throw or exit the program (modulo assertion failures).
+static bool isReadOnlyConstantEvaluableCall(FullApplySite applySite) {
+  assert(applySite);
+  SILFunction *callee = applySite.getCalleeFunction();
+  if (!callee || !isConstantEvaluable(callee)) {
+    return false;
+  }
+  // Here all effects of the call is restricted to its indirect results, which
+  // must have value semantics. If there are no indirect results, the call must
+  // be read-only.
+  return applySite.getNumIndirectSILResults() == 0;
+}
+
+/// A scope-affecting instruction is an instruction which may end the scope of
+/// its operand or may produce scoped results that require cleaning up. E.g.
+/// begin_borrow, begin_access, copy_value, a call that produces a owned value
+/// are scoped instructions. The scope of the results of the first two
+/// instructions end with an end_borrow/acess instruction, while those of the
+/// latter two end with a consuming operation like destroy_value instruction.
+/// These instruction may also end the scope of its operand e.g. a call could
+/// consume owned arguments thereby ending its scope. Dead-code eliminating a
+/// scope-affecting instruction requires fixing the lifetime of the non-trivial
+/// operands of the instruction and requires cleaning up the end-of-scope uses
+/// of non-trivial results.
+///
+/// \param inst instruction that checked for liveness.
+/// \param knownPureValueTypes a set of known non-trivial types that are known
+/// to have pure value value semantics. These types are populated when a
+/// constant_evaluable call is seen, as it would accept only types with value
+/// semantics.
+static bool
+isScopeAffectingInstructionDead(SILInstruction *inst,
+                                SmallPtrSetImpl<CanType> &knownPureValueTypes) {
+  SILFunction *fun = inst->getFunction();
+  assert(fun && "Instruction has no function.");
+  // Only support ownership SIL for scoped instructions.
+  if (!fun->hasOwnership()) {
+    return false;
+  }
+  // If the instruction has any use other than end of scope use or destroy_value
+  // use, bail out.
+  if (!hasOnlyEndOfScopeOrDestroyUses(inst)) {
+    return false;
+  }
+  // If inst is a copy or beginning of scope, inst is dead, since we know that
+  // it is used only in a destroy_value or end-of-scope instruction.
+  if (getSingleValueCopyOrCast(inst))
+    return true;
+
+  switch (inst->getKind()) {
+  case SILInstructionKind::LoadInst: {
+    LoadOwnershipQualifier loadOwnershipQual =
+        cast<LoadInst>(inst)->getOwnershipQualifier();
+    // If the load creates a copy, it is dead, since we know that if at all it
+    // is used, it is only in a destroy_value instruction.0
+    return (loadOwnershipQual == LoadOwnershipQualifier::Copy ||
+            loadOwnershipQual == LoadOwnershipQualifier::Trivial);
```

Comment:
Consider handling `load [take]` (or add a FIXME). Just insert a `destroy_addr` in its place.

Reply:
Also comment that we should handle `load_borrow` here`